
# Spruce-cold-stress

Here guides used in Spruce's transcriptomics data analysis are included.

This directory contains Biological Quality Analysis (BiologicalQA) and Differential Expression (DE) scripts for each tissue (needles and roots) and details about the steps followed to analyze the publication data.  

RNA-Seq pipeline and accesory code like plot.multidensity.R script which is necessary for BiologicalQA analysis deveolped by Bioinformatics UPSC groups is avilable here:
https://github.com/UPSCb/UPSCb

A detailed version of the workflow is available on Wiki:
https://github.com/avergro/Spruce-cold-stress/wiki.


# Running title: Surviving winter- the Norway spruce cold transcriptome

Title: Norway spruce deploys canonical and delayed cold acclimation responses in a  tissue-specific manner. 
 
Authors 
	Alexander Vergara1*, Julia Haas2*, Paulina Stachula2, Nathaniel Street2 and Vaughan Hurry1

# Affiliations	
	1Umeå Plant Science Centre, Department of Forest Genetics and Plant Physiology, Swedish University of Agricultural Sciences, SE901 83 Umeå, Sweden. 2Umeå Plant Science Centre, Department of Plant Physiology, Umeå University, SE901 87 Umeå, Sweden.

# Abstract

Cold acclimation in plants is a complex phenomenon that involves a number of different stress-responsive transcriptional and metabolic pathways.  Most gene expression studies have addressed short-term cold acclimation responses in herbaceous plants while few have focused on long-term cold responses, and even fewer have investigated perennial evergreens such as conifers, which survive extremely low temperatures during winter. To characterize the transcriptome changes during cold acclimation in Picea abies (L.) H. Karst (Norway spruce), we performed RNA-Sequencing analysis of needles and roots subjected to cold treatments through a chilling progression (5 °C) until 10 days at freeze temperature (-5 °C). Comparing gene expression responses of needles against Arabidopsis thaliana L. (Arabidopsis) leaves, our results showed that early transient inductions were observed in both species but the transcriptional response of Norway spruce was delayed. Our results indicate that, similar to herbaceous species, Norway spruce principally utilizes early response transcription factors (TF) that belong to the AP2/ERF superfamily and NACs. However, unique to the Norway spruce response was a large group of TFs that mounted a late transcriptional response to low temperature. A regulatory network analysis identified key TFs, including a root-specific bHLH101-like and other members of the same family with a pervasive role in cold regulation, such as ICE1 and AKS3-like and also members of the NAC (anac47 and anac28-like) and AP2/ERF superfamilies (DREB2 and ERF3-like), providing new functional insights into cold stress response strategies in Norway spruce.

# Data availability

Sequencing data are available at the European Nucleotide Archive (ENA) as accession PRJEB26934.

# References

http://www.epigenesys.eu/en/protocols/bio-informatics/1283-guidelines-for-rna-seq-data-analysis

