# Spruce-cold-stress

Here guides used in Spruce's transcriptomics data analysis are included.

This directory contains Biological Quality Analysis (BiologicalQA) and Differential Expression (DE) scripts for each tissue (needles and roots) and details about the steps followed to analyze the publication data.  

RNA-Seq pipeline and accesory code necessary for BiologicalQA analysis deveolped by Bioinformatics UPSC group is avilable here:

https://github.com/UPSCb/UPSCb

A detailed version of the workflow is available on Wiki:

https://github.com/avergro/Spruce-cold-stress/wiki/Spruce-cold-stress-wiki


### Running title: 
## Norway spruce deploys tissue-specific cold acclimation responses


Authors 
	Alexander Vergara1*, Julia Haas2*, Tuuli Aro1, Paulina Stachula2, Nathaniel Street2 and Vaughan Hurry1

### Affiliations	
1. Umeå Plant Science Centre, Department of Forest Genetics and Plant Physiology, Swedish University of Agricultural Sciences, SE901 83 Umeå, Sweden. 
2. Umeå Plant Science Centre, Department of Plant Physiology, Umeå University, SE901 87 Umeå, Sweden.

### Short Title: 
### Surviving winter: the Norway spruce cold transcriptome

 
## Abstract

Cold acclimation in plants is a complex phenomenon involving numerous stress-responsive transcriptional and metabolic pathways. Existing gene expression studies have primarily addressed cold acclimation responses in herbaceous plants, few have focused on perennial evergreens, such as conifers, that survive extremely low temperatures during winter. To characterize the transcriptome changes during cold acclimation in Picea abies (L.) H. Karst (Norway spruce), we performed RNA-Sequencing analysis of needles and roots subjected to a chilling progression (5 °C) followed by 10 days at freezing temperature (-5 °C). Comparing gene expression responses of needles against Arabidopsis thaliana L. (Arabidopsis) leaves, our results showed that early transient inductions were observed in both species, but the main transcriptional response of Norway spruce was delayed and was associated with slower cold acclimation. Despite this difference in timing, our results indicate that, similar to herbaceous species, Norway spruce principally utilizes early response transcription factors (TFs) of the APETALA 2/ethylene-responsive element binding factor (AP2/ERF) superfamily and NAM (no apical meristem)/ATAF (Arabidopsis Transcription Factors)/CUC (cup shaped cotyledon) (NACs). However, unique to the Norway spruce response was a large group of TFs that mounted the late transcriptional response to low temperature. Regulatory network analysis identified key conserved TFs, including a root-specific bHLH101 homolog and other members of the same family with a pervasive role in cold regulation, such as homologs of ICE1 and AKS3 and also homologs of the NAC (anac47 and anac28) and AP2/ERF superfamilies (DREB2 and ERF3), providing new functional insights into cold stress response strategies in Norway spruce.

## Data availability

Sequencing data are available at the European Nucleotide Archive (ENA) as accession PRJEB26934.

## References

Delhomme et. al. (2014) Guidelines for RNA-Seq data analysis. EpyGeneSys Protocol (prot 67). https://http://www.epigenesys.eu/en/protocols/bio-informatics/1283-guidelines-for-rna-seq-data-analysis
